torch/csrc/autograd/python_variable_indexing.cpp

```diff
@@ -376,44 +323,49 @@ int THPVariable_setitem(PyObject* self, PyObject* index, PyObject* py_value) {
   } else {
     value = valueToTensor(self_.options(), py_value);
   }
+  bool is_tracing = torch::jit::tracer::isTracing();
 
   // handle simple types: integers, slices, ellipsis, bool
-  if (index == Py_False) { // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
-    // do nothing for false (technically we should check the size, but we don't have
-    // real 0-sized shapes.
-    return 0;
-  } else if (index == Py_Ellipsis) {
-    copy_to(self_, value);
-    return 0;
-  } else if (index == Py_None || index == Py_True) { // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
-    copy_to(self_.unsqueeze(0), value);
-    return 0;
-  } else if (THPUtils_checkLong(index)) {
-    copy_to(applySelect(self_, 0, index), value);
-    return 0;
-  } else if (PySlice_Check(index)) {
-    copy_to(applySlice(self_, 0, index), value);
+  if (is_tracing) {
+    if (THPUtils_checkLong(index) && THPVariable_Check(index)) {
+      recordSelectTrace(THPVariable_Unpack(index));
+    } else if (PySlice_Check(index)) {
+      Tensor start_tensor, stop_tensor, step_tensor;
+      extractTensorsFromSlice(index, start_tensor, stop_tensor, step_tensor);
+      recordSliceTrace(start_tensor, stop_tensor, step_tensor);
+    }
+  }
+  if (index == Py_False || index == Py_Ellipsis || index == Py_None || index == Py_True || THPUtils_checkLong(index) || PySlice_Check(index)) {
+    indexToTensorIndex(self_, index, at::indexing::tensor_index_tmp);
```

Comment:
There's no reason to put the temporary in global. Just stick it as a stack local variable in this function... and please stop using mutable references LOL

Reply:
BTW, you don't even need to reuse the struct because if you implement indexToTensorIndex properly you should get return value optimization, meaning that TensorIndex gets constructed directly in the parent stack frame